src/m_gsm_symbolic/sofie_test_parser.py

```diff
@@ -0,0 +1,225 @@
+    "sample": lambda items, n = 2: random.sample(items, n), # trying to make the dict handle multiple items 
+    "divides": lambda a, b: a % b == 0,}
+
+
+def eval_braces(match, num_values):
+    """ Helper function to evalulate the expressions inside the {}"""
+    expr = match.group(1)  # Get the expression inside the {}
+    try:
+        result = eval(expr, {}, num_values)  # Evaluate the math
+        return str(result)  # Replace with the result
+    except Exception as e:
+        return match.group(0)
+
+
+def eval_expressions(text: str, new_values: dict[str, str]) -> str:
+    '''
+    Evaluate all expressions like {age1+age_diff} and <<expression=result>> in the text.
+    Only keep numeric values. Uses the eval_braces helper function to replace expressions inside brace
+    '''
+    num_values = {}
+    for i, v in new_values.items():
+        try:
+            num_values[i] = int(v) # Try converting the value to an integer
+        except ValueError:
+            pass  # Skip non-numeric values like names    
+    text = re.sub(r"\{([^,{}][^{}]*)\}", lambda match: eval_braces(match, num_values), text)
+    return text
+
+
+@dataclass
+class GeneratedQuestion:
+    question: str
+    answer: str
+    id_orig: int
+    id_shuffled: int
+
+
+@dataclass
+class AnnotatedQuestion:
+    question: str
+    answer: str
+    id_orig: int
+    id_shuffled: int
+    question_annotated: str
+    answer_annotated: str
+
+    @classmethod
+    def from_json(cls, json_data: str) -> Self:
+        data = json.loads(json_data)
+        return cls(**data)
+
+    @property
+    def init(self) -> str:
+        """extract init section from question_annotated"""
+        return self.question_annotated.split("#init:")[1].split("#conditions:")[0]
+
+    @property
+    def default_values(self) -> dict[str, str]:
+        """extract default values from both the annotated question and answer"""
+        default_values = {}
+
+        # find {var_name, value} pairs
+        regex = r"\{(\w+),(\w+)\}" # any alphanumeric chr
+
+        # hmmm might not be the smartest solution? But I couldnt replace variable values for the 
+        # answer annoted, if I didnt extract the default dict from it? 
+        annotated_question_answer = self.question_annotated + self.answer_annotated
+
+        matches = re.findall(regex, annotated_question_answer) # returns a list containing all matches
+        for var_name, value in matches:
+            default_values[var_name] = value
+        return default_values
+
+    @staticmethod
+    def remove_calculations(text: str) -> str:
+        """remove calculations from the text. re.sub() "replaces" the calculations with noting"""
+        return re.sub(r"<<.*?>>", "", text)  
+    
+    def apply_init_rules(self,
+                        names: list[str],
+                        multiple_ice: list[int],
+                        multi_times: list[int],
```

Comment:
We should start generalizing to more than just this one question

Reply:
A way to do it could be:

```
def apply_init_rules(self, replacements: dict[str, Any] = {}):
```
where the dict would be e.g. {"names": ["Ane"]}

 If a key is not present then it can just draw from a predefined list in the script.